src/block_coordinate_algorithms.jl

```diff
@@ -0,0 +1,355 @@
+            end
+        end
+
+        #####################
+
+        first_iter = false
+
+        # Update all dimensions simulatenously
+        if update_order == full
+
+            if momentum === nothing || first_iter
+                grad!(gradient, x)
+                if momentum !== nothing
+                    gtemp .= gradient
+                end
+            else
+                grad!(gtemp, x)
+                @memory_mode(memory_mode, gradient = (momentum * gradient) + (1 - momentum) * gtemp)
+            end
+
+            v = cat(compute_extreme_point(lmo, tuple([selectdim(gradient, ndim, i) for i=1:l]...))..., dims = ndim)
+            d = muladd_memory_mode(memory_mode, d, x, v)
+            gamma = perform_line_search(
+                line_search,
+                t,
+                (x) -> f(x),
+                (storage, x) -> grad!(storage, x),
+                gradient,
+                x,
+                d,
+                1.0,
+                linesearch_workspace,
+                memory_mode,
+            )
+
+            dual_gap = sum(x.*gradient) - sum(v.*gradient)
+    
+            x = muladd_memory_mode(memory_mode, x, gamma, d)
+        else
+            for j=1:l
+
+                if momentum === nothing || first_iter
+                    grad!(gradient, x)
+                    if momentum !== nothing
+                        gtemp .= gradient
+                    end
+                else
+                    grad!(gtemp, x)
+                    @memory_mode(memory_mode, gradient = (momentum * gradient) + (1 - momentum) * gtemp)
+                end
+                    
+                if update_order == cyclic
+                    i = j
+                elseif update_order == stochastic
+                    i = rand(1:l)
+                elseif update_order == progressive
+                    weights = progress ./sum(progress)
+                    i = findfirst(cumsum(weights) .>= rand())
+                else
+                    @warn "Unknown update_order: $(update_order)"
+                end
+                
+                
+                v = copy(x) # This is equivalent to setting the rest of d to zero
+                indices = [idx < ndim ? Colon() : i for idx=1:ndim]
+                v[indices...] = compute_extreme_point(lmo.lmos[i], gradient[indices...])
+
+                dgi = fast_dot(x[indices...], gradient[indices...]) - fast_dot(v[indices...], gradient[indices...])
+
+                if dual_gap != Inf
+                    progress[i] = copy(dual_gaps[i]) - dgi
+                end
+                
+                dual_gaps[i] = dgi
+                dual_gap = sum(dual_gaps)
+
+                d = muladd_memory_mode(memory_mode, d, x, v)
+                gamma = perform_line_search(
+                    line_search,
+                    t,
+                    (x) -> f(x),
+                    (storage, x) -> grad!(storage, x),
+                    gradient,
```

Comment:
```suggestion
                    f,
                    grad!,
                    gradient,
```
this is equivalent without creating an additional anonymous function